tools/image-scan

```diff
@@ -1,7 +1,191 @@
-#! /bin/bash -eu
-echo "=========== ECR Docker Image Vulnerability Scan for $IMAGE_ID"
-cd  $JUPYTERHUB_DIR
-image-scan-report "$IMAGE_UBUNTU_NAME" $IMAGE_VULNERABILITY_LEVEL >ecr-scan-report.yaml
+#! /usr/bin/env python
+# -*-python-*-
 
-echo "=========== ECR Scan Summary ================================"
-image-scan-summary ecr-scan-report.yaml
+"""This tool pulls down Docker image scan reports from ECR based on
+the current environment which is used to identify the image.
+
+Run "image-scan-report --help" for usage information.
+
+The result of the report is a YAML file limited to CVE's which are
+at the minimum severity level or greater.
+
+For each CVE, the tool follows the URI reported by ECR (to an Ubuntu
+website) and extracts the CVE status from that site, e.g. "Deferred" or
+"Needed".
+"""
+
+import sys
+import os
+import subprocess
+import json
+import yaml
+import copy
+import argparse
+import requests
+import bs4
+import time
+
+
+KEEP_LEVELS = {
+    "CRITICAL": ["CRITICAL"],
+    "HIGH": ["CRITICAL", "HIGH"],
+    "MEDIUM": ["CRITICAL", "HIGH", "MEDIUM"],
+    "LOW": ["CRITICAL", "HIGH", "MEDIUM", "LOW"],
+    "INFORMATIONAL": ["CRITICAL", "HIGH", "MEDIUM", "LOW", "INFORMATIONAL"],
+    "ALL": ["CRITICAL", "HIGH", "MEDIUM", "LOW", "INFORMATIONAL"],
+}
+
+FAIL_LEVELS = ["CRITCAL", "HIGH"]
+
+
+def run(cmd, cwd="."):
+    result = subprocess.run(
+        cmd.split(),
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
+        check=True,
+        cwd=cwd,
+    )  # maybe succeeds
+    return result.stdout
+
+
+def _get_scan_results(image_digest, image_tag):
+    """Issue an AWS command to dump the ECR scan results as JSON, load the
+    JSON,  and return the resulting dict.
+    """
+    admin_arn = os.environ["ADMIN_ARN"]
+    image_repo = os.environ["IMAGE_REPO"]
+    ecr_account_to_use = os.environ["ECR_ACCOUNT_ID"]
+
+    image_id_arg = f"imageTag={image_tag}"
+    if image_digest:
+        image_id_arg = f"imageDigest={image_digest}"
+
+    print(
+        f"Fetching ECR vulnerability scan for registry={ecr_account_to_use} repo={image_repo} tag={image_tag}",
+        file=sys.stderr,
+    )
+
+    scan_results = run(
+        f"awsudo -d 3600 {admin_arn}  aws ecr describe-image-scan-findings "
+        f"--no-paginate "
+        f"--registry-id {ecr_account_to_use} "
+        f"--repository-name {image_repo} "
+        f"--image-id {image_id_arg}"
+    )
+    return json.loads(scan_results)
+
+
+def get_scan_results(image_digest):
+    image_tag = os.environ["IMAGE_TAG"]
+    try:
+        return _get_scan_results(image_digest, image_tag)
+    except Exception:
+        print(
+            f"Failed fetching {image_tag}.  Trying for approved tag.", file=sys.stderr
+        )
+        if image_tag.startswith("unscanned-"):
+            return _get_scan_results(image_digest, image_tag[len("unscanned-") :])
```

Comment:
I must be missing something here.  How can we get results for an image that has not been scanned (to completion at least)?

Reply:
You're right,  it's possible this is just a logic error and ill-conceived  response to the error which started occurring when we started using unscanned-;  IMAGE_TAG basically became wrong in this context.

It's not out-of-the-question to drop the script altogether but IMHO it's still good to be able to fetch the results ourselves and correlate them to Ubuntu...  even though Ubuntu can never do anything about the results.

If you want to clean it up more post-merge,  have at it.

Reply:
One other possible thing:  if the scan fails,  it is never retagged.

Reply:
For sure.  It's definitely worth having around.  I think it's fine, I was just staring at it and wondering how that was supposed to be handled :+1: